milli/src/update/word_prefix_pair_proximity_docids.rs

```diff
@@ -169,78 +373,549 @@ impl<'t, 'u, 'i> WordPrefixPairProximityDocids<'t, 'u, 'i> {
+// writer don't conflict with values in the database.
+pub fn write_into_lmdb_database_without_merging(
+    wtxn: &mut heed::RwTxn,
+    database: heed::PolyDatabase,
+    writer: grenad::Writer<std::fs::File>,
 ) -> Result<()> {
-    *current_prefixes = match current_prefixes.take() {
-        Some(prefixes) if w2.starts_with(&prefixes[0]) => Some(prefixes),
-        _otherwise => {
-            write_prefixes_in_sorter(prefixes_cache, word_prefix_pair_proximity_docids_sorter)?;
-            prefix_fst_keys.iter().find(|prefixes| w2.starts_with(&prefixes[0]))
+    let file = writer.into_inner()?;
+    let reader = grenad::Reader::new(BufReader::new(file))?;
+    if database.is_empty(wtxn)? {
+        let mut out_iter = database.iter_mut::<_, ByteSlice, ByteSlice>(wtxn)?;
+        let mut cursor = reader.into_cursor()?;
+        while let Some((k, v)) = cursor.move_on_next()? {
+            // safety: we don't keep references from inside the LMDB database.
+            unsafe { out_iter.append(k, v)? };
         }
-    };
+    } else {
+        let mut cursor = reader.into_cursor()?;
+        while let Some((k, v)) = cursor.move_on_next()? {
+            database.put::<_, ByteSlice, ByteSlice>(wtxn, k, v)?;
+        }
+    }
+    Ok(())
+}
 
-    if let Some(prefixes) = current_prefixes {
-        buffer.clear();
-        buffer.extend_from_slice(w1.as_bytes());
-        buffer.push(0);
-        for prefix in prefixes.iter() {
-            if prefix.len() <= max_prefix_length && w2.starts_with(prefix) {
-                buffer.truncate(w1.len() + 1);
-                buffer.extend_from_slice(prefix.as_bytes());
-                buffer.push(prox);
-
-                match prefixes_cache.get_mut(buffer.as_slice()) {
-                    Some(value) => value.push(data.to_owned()),
-                    None => {
-                        prefixes_cache.insert(buffer.clone(), vec![data.to_owned()]);
+struct Allocations {
+    byte_vectors: Vec<Vec<u8>>,
+    mergeable_data_vectors: Vec<Vec<Cow<'static, [u8]>>>,
+}
+impl Default for Allocations {
+    fn default() -> Self {
+        Self {
+            byte_vectors: Vec::with_capacity(65_536),
+            mergeable_data_vectors: Vec::with_capacity(4096),
+        }
+    }
+}
+impl Allocations {
+    fn take_byte_vector(&mut self) -> Vec<u8> {
+        self.byte_vectors.pop().unwrap_or_else(|| Vec::with_capacity(16))
+    }
+    fn take_mergeable_data_vector(&mut self) -> Vec<Cow<'static, [u8]>> {
+        self.mergeable_data_vectors.pop().unwrap_or_else(|| Vec::with_capacity(8))
+    }
+
+    fn reclaim_byte_vector(&mut self, mut data: Vec<u8>) {
+        data.clear();
+        self.byte_vectors.push(data);
+    }
+    fn reclaim_mergeable_data_vector(&mut self, mut data: Vec<Cow<'static, [u8]>>) {
+        data.clear();
+        self.mergeable_data_vectors.push(data);
+    }
+}
+
+#[derive(Default, Debug)]
+struct PrefixTrieNode {
```

Comment:
Would it be possible to explain how this struct works, with a simple doc comment, please?

Reply:
done 👍  let me know if the explanation is good enough